py/server/deephaven/experimental/iceberg.py

```diff
@@ -0,0 +1,251 @@
+        except Exception as e:
+            raise DHError(e, "Failed to build Iceberg instructions") from e
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self._j_object
+
+
+class IcebergCatalogAdapter(JObjectWrapper):
+    """
+    This class provides an interface for interacting with Iceberg catalogs. It allows listing namespaces, tables and
+    snapshots, as well as reading Iceberg tables into Deephaven tables.
+    """
+    j_object_type = _JIcebergCatalogAdapter or type(None)
+
+    def __init__(self, j_object: _JIcebergCatalogAdapter):
+        self.j_catalog_adapter = j_object
+
+    def namespaces(self, namespace: Optional[str] = None) -> Table:
+        """
+        Returns the namespaces in the catalog as a Deephaven table.
+
+        Args:
+            namespace (Optional[str]): the higher-level namespace from which to list namespaces; if omitted, the
+                top-level namespaces are listed.
+
+        Returns:
+            a table containing the namespaces.
+        """
+
+        if namespace is not None:
+            return Table(self.j_object.listNamespaces(namespace))
+        return Table(self.j_object.listNamespacesAsTable())
+
+    def tables(self, namespace: str) -> Table:
+        """
+        Returns the list of tables in the provided namespace as a Deephaven table.
+
+        Args:
+            namespace (str): the namespace from which to list tables.
+
+        Returns:
+            a table containing the tables in the provided namespace.
+        """
+
+        if namespace is not None:
+            return Table(self.j_object.listTablesAsTable(namespace))
+        return Table(self.j_object.listTablesAsTable())
+
+    def snapshots(self, table_identifier: str) -> Table:
+        """
+        Returns the list of snapshots of the provided table as a Deephaven table.
+
+        Args:
+            table_identifier (str): the table from which to list snapshots.
+
+        Returns:
+            a table containing the snapshot information.
+        """
+
+        return self.j_object.listSnapshotsAsTable(table_identifier)
+
+    def read_table(self, table_identifier: str, instructions: Optional[IcebergInstructions] = None, snapshot_id: Optional[int] = None) -> Table:
+        """
+        Reads the table from the catalog using the provided instructions. Optionally, a snapshot id can be provided to
+        read a specific snapshot of the table.
+
+        Args:
+            table_identifier (str): the table to read.
+            instructions (Optional[IcebergInstructions]): the instructions for reading the table. These instructions
+                can include column renames, table definition, and specific data instructions for reading the data files
+                from the provider. If omitted, the table will be read with default instructions.
+            snapshot_id (Optional[int]): the snapshot id to read; if omitted the most recent snapshot will be selected.
```

Comment:
Where would they get a list of valid snapshot ids?

Reply:
One option is by calling `snapshots(table_id)` and providing a value from the table. The user may also have knowledge the the snapshots from the external database.